### ISSUE TITLE
Request for Evaluation Code and Test Sample Data

### ISSUE THREAD

**Issue by PengchengShi1220** [ImranNust/AortaSeg24#2]
Dear organizer,

I have successfully tested my model using both the local Docker environment and the Grand Challenge Algorithms platform, achieving an average Dice score of over 0.70 across 23 vascular categories on the validation set, where the prediction appears consistent with the ground truth in ITK-SNAP 3D visualization. However, there is a significant drop in the Dice score at the submission stage, which is perplexing.
![image](https://github.com/user-attachments/assets/22c53e12-4470-4ea1-beee-07839670af88)
![image](https://github.com/user-attachments/assets/03f36558-fc39-4390-b0ec-c24dfd0f41c7)
![image](https://github.com/user-attachments/assets/a76e16b8-81f0-4760-a64a-d76cfdd9a45c)

Could you please make the evaluation code available and consider releasing one example from the test set (only the imaging data) to allow participants to verify consistency across the training and test sets on the Algorithms platform?

Best regards,
Pengcheng

**Comment by PengchengShi1220** [comment]
I have downloaded the training data and checked them using ITK-SNAP, and they are consistently in RAI orientation. Despite redownloading them, they remain in RAI orientation. I would like to clarify one crucial point: Is the orientation of the test data consistent with that of the training data? 
![image](https://github.com/user-attachments/assets/043c8ebc-2d6b-47ae-b33f-a05b357f0f12)

All training data is oriented in RAI, and our model (based on nnU-Net framework) has been trained in RAI. Upon reviewing the code in your repository, specifically the scripts [training/utils/dataset.py](https://github.com/ImranNust/AortaSeg24/blob/main/training/utils/dataset.py) and [validation/utils/dataset.py](https://github.com/ImranNust/AortaSeg24/blob/main/validation/utils/dataset.py), we noticed that there is a step `transformsOrientationd(keys=["image", "label"], axcodes="RAS")`.

However, this transformation step seems to be absent in the inference script [Docker_Preparation_For_Submission/inference.py](https://github.com/ImranNust/AortaSeg24/blob/main/Docker_Preparation_For_Submission/inference.py). Could it be possible that the test data orientation was converted to RAS, or might there be another reason for this discrepancy?

**Comment by ImranNust** [comment]
Hi,

The evaluation code used in the background on the Grand-Challenge website has been shared with you. You can access it via this [link](https://github.com/ImranNust/AortaSeg24/tree/main/evaluation_docker_for_validation_phase).

Regarding the answers to your questions:



- **Training Data**: When viewed in ITK-Snap, the training data is in the RAI orientation. In 3D Slicer, the direction matrix is:
  ```math
  \begin{bmatrix}
  -1 & 0 & 0 \\
  0 & -1 & 0 \\
  0 & 0 & 1
  \end{bmatrix}
  ```

- **Testing Data**: When viewed in ITK-Snap, the testing data is in the LPI orientation. In 3D Slicer, the direction matrix is:
  ```math
  \begin{bmatrix}
  1 & 0 & 0 \\
  0 & 1 & 0 \\
  0 & 0 & 1
  \end{bmatrix}
  ```

For visualization purposes, we use 3D Slicer.

If you need further assistance, we can arrange a Zoom meeting. Please email me at a.cosman@ufl.edu to schedule an online meeting.

**Comment by PengchengShi1220** [comment]
Hi,

Thank you for sharing the evaluation code. I have tested one of the validation data subject009_CTA.mha using the same Docker image on the Algorithm platform. The results are as shown in the image: ![image](https://github.com/user-attachments/assets/c1362e16-f111-49fe-a5a8-2a1c1dc7712e). I also computed the Dice score locally using the [evaluation.py](https://github.com/ImranNust/AortaSeg24/blob/main/evaluation_docker_for_validation_phase/evaluate.py) script on subject009's predictions and ground truth, and obtained an Average Dice Score of 0.747, which is in line with expectations and significantly higher than 0.05. This indicates that the evaluation code is functioning correctly.
![image](https://github.com/user-attachments/assets/50c77992-1fa0-4111-a26b-fe9d5550c6b0)

Could you please consider releasing one example from the test set (just the imaging data) to allow participants to verify consistency across the training and test sets on the Algorithms platform? The current situation is quite confusing, and it seems likely that there might be inconsistencies between the training and test data.

Thank you for your assistance.

**Comment by ImranNust** [comment]
Are you using "transformsOrientationd(keys=["image", "label"], axcodes="RAS")" in your inference file ?

**Comment by PengchengShi1220** [comment]
Hi,

No, I am not using `transformsOrientationd(keys=["image", "label"], axcodes="RAS")`. My model, based on the nnU-Net framework, was trained on the original RAI data. As long as the test data is also RAI, I obtain correct results without converting to RAS. I don't see the need for this conversion in this context. Given the limited number of submissions, I'm hesitant to attempt this change without a clear reason.

**Comment by ImranNust** [comment]
You need to ensure that the segmentation output produced by your model retains the same properties as the input 3D volume used to generate that segmentation. For example, if the origin of the input 3D volume is (0,0,0), then the origin of the output segmentation should also be (0,0,0). Similarly, if the direction matrix of the input is (1, 0, 0; 0, 1, 0; 0, 0, 1), then the direction matrix of the output should be (1, 0, 0; 0, 1, 0; 0, 0, 1). All other properties should match as well.

You can achieve this by using the following commands:

```python
image.SetDirection(direction)  # Set the direction matrix
image.SetOrigin(origin)        # Set the origin
```

These commands ensure that the output segmentation aligns correctly with the input 3D volume.

**Comment by PengchengShi1220** [comment]
Thank you for the reminder. I have already addressed this in my nnU-Net version of the inference file. Please refer to my [inference.py](https://github.com/PengchengShi1220/AortaSeg24/blob/main/inference.py). I ensure the consistency of header information between the input and output files using the following code:

```python
image = SimpleITK.GetImageFromArray(pred_array)
image.SetDirection(props['sitk_stuff']['direction'])
image.SetOrigin(props['sitk_stuff']['origin'])
image.SetSpacing(props['sitk_stuff']['spacing'])
```

**Comment by ImranNust** [comment]
The training data was originally in the LAI coordinate system with the following direction matrix:

```math
[-1, 0, 1; \\
0, -1, 0; \\
0, 0, 1]
```

For this reason, during training, we applied the transformation:

```python
transformsOrientationd(keys=["image", "label"], axcodes="RAS")
```

However, to streamline the inference process, the test data was already provided in the RAS format with this direction matrix:

```math
[1, 0, 0; \\
0, 1, 0; \\
0, 0, 1]
```

Consequently, the transformation:

```python
transformsOrientationd(keys=["image", "label"], axcodes="RAS")
```

is not needed for the test data. Rest assured, the test dataset maintains the same properties as the training data.


This version clarifies the reasoning and ensures consistency in terminology and formatting.

**Comment by PengchengShi1220** [comment]
Hi,

Thank you for the information. Could you please clarify where the information about the training data originally being in the LAI coordinate system and the test data being provided in the RAS format comes from? I used ITK-SNAP to check the downloaded training data (dated April 12, 2024), and the orientation was RAI. I'm confused about the LAI and RAS details you mentioned.

![image](https://github.com/user-attachments/assets/eade795f-c5b9-466c-9fbd-266c9aeafda4)

**Comment by PengchengShi1220** [comment]
Hi,

I want to confirm that RAI corresponds to the direction matrix [1,0,0; 0,1,0; 0,0,1], while RAS corresponds to [1,0,0; 0,1,0; 0,0,-1]. So, you mean that apart from the orientation, the training data and test data are consistent. If I train a model based on RAS training data and perform inference on the test data, it should resolve the issue, correct? Given the limited number of submissions, I hope you understand my need for clarity.

**Comment by ImranNust** [comment]
I apologize for any confusion. Let me clarify the details for you:

- **Training Data**: When viewed in ITK-Snap, the training data is in the RAI orientation. In 3D Slicer, the direction matrix is:
  ```math
  \begin{bmatrix}
  -1 & 0 & 0 \\
  0 & -1 & 0 \\
  0 & 0 & 1
  \end{bmatrix}
  ```

- **Testing Data**: When viewed in ITK-Snap, the testing data is in the LPI orientation. In 3D Slicer, the direction matrix is:
  ```math
  \begin{bmatrix}
  1 & 0 & 0 \\
  0 & 1 & 0 \\
  0 & 0 & 1
  \end{bmatrix}
  ```

Other than the orientation, all properties of the testing and training data are the same. I hope this information helps you modify your inference code to produce accurate results.

My question is: will you convert your training data from RAI to LPI and then retrain your model, or will you adapt your inference code accordingly?

Additionally, we have increased the submission limit from five to ten, and the final submission date has been extended to August 15, 2024. I hope this helps.

**Comment by PengchengShi1220** [comment]
Thank you very much for the prompt clarification. I appreciate the increased submission limit and extended deadline. I will first try to adapt my inference code to handle the LPI orientation. If that approach fails, I will then convert the training data from RAI to LPI (ITK-SNAP) before retraining my model.

Best regards,
Pengcheng

**Comment by ImranNust** [comment]
We appreciate your contribution and look forward to see your results soon.

**Comment by Puzzled-Hui** [comment]
> The training data was originally in the LAI coordinate system with the following direction matrix:
> 
> [
> −
> 1
> ,
> 0
> ,
> 1
> ;
> 0
> ,
> −
> 1
> ,
> 0
> ;
> 0
> ,
> 0
> ,
> 1
> ]
> For this reason, during training, we applied the transformation:
> 
> ```python
> transformsOrientationd(keys=["image", "label"], axcodes="RAS")
> ```
> 
> However, to streamline the inference process, the test data was already provided in the RAS format with this direction matrix:
> 
> [
> 1
> ,
> 0
> ,
> 0
> ;
> 0
> ,
> 1
> ,
> 0
> ;
> 0
> ,
> 0
> ,
> 1
> ]
> Consequently, the transformation:
> 
> ```python
> transformsOrientationd(keys=["image", "label"], axcodes="RAS")
> ```
> 
> is not needed for the test data. Rest assured, the test dataset maintains the same properties as the training data.
> 
> This version clarifies the reasoning and ensures consistency in terminology and formatting.

Training Data in LAI and Testing Data in RAS?

**Comment by ImranNust** [comment]
I apologize for any confusion. Let me clarify the details for you:

- **Training Data**: When viewed in ITK-Snap, the training data is in the RAI orientation. In 3D Slicer, the direction matrix is:
  ```math
  \begin{bmatrix}
  -1 & 0 & 0 \\
  0 & -1 & 0 \\
  0 & 0 & 1
  \end{bmatrix}
  ```

- **Testing Data**: When viewed in ITK-Snap, the testing data is in the LPI orientation. In 3D Slicer, the direction matrix is:
  ```math
  \begin{bmatrix}
  1 & 0 & 0 \\
  0 & 1 & 0 \\
  0 & 0 & 1
  \end{bmatrix}
  ```

I hope this information helps you modify your inference code to produce accurate results.

**Comment by ImranNust** [comment]
The discussion at https://grand-challenge.org/forums/forum/multi-class-segmentation-of-aortic-branches-and-zones-in-cta-726/topic/executing-failed-2358/ can be helpful.

**Comment by PengchengShi1220** [comment]
The code to resolve the issue can be referenced here: [AortaSeg24 inference](https://github.com/PengchengShi1220/AortaSeg24/blob/main/inference.py). Closing the issue.
